### ISSUE TITLE
Massive amount of time spent in "Compiling schema"

### ISSUE THREAD

**Issue by drewschaeffer** [cube-js/cube#7893]
**Problem**

Looking into performance of our CubeJs I notice that we often have massive amounts of time spent in schema compile. We are using dynamic schema generation utilizing asyncModel with a refresh_key of 1 hour. We are also running a large number of cube instances compared to our load.

I'm curious what are some strategies for us to investigate this performance issue or possibly optimize the configuration of our Cube?

For example, here is one query that took 59 seconds in total where over 54 seconds of that was spent in schema compile. Any suggestions are appreciated!

<img width="821" alt="image" src="https://github.com/cube-js/cube/assets/6683917/d3049e35-7be1-46d4-a160-abefe613cb5f">

**Comment by igorlukanin** [comment]
Hi @drewschaeffer 👋 

> We are using dynamic schema generation utilizing asyncModel

I would look into what is happening within your `asyncModule`. Most likely, you make network calls there—right?—and that definitely contributes to the overall data model compilation time.

Also, it would be great to check your [`schema_version`](https://cube.dev/docs/reference/configuration/config#schema_version) implementation—how efficient is that?

Finally, I have no insight into whether you use multitenancy or not. Cube maintains a separate compiled data model for each tenant, and it would recompile it for each tenant when schema version changes.

Also, just FYI, if you run Cube in Cube Cloud, you can use [Performance Insights](https://cube.dev/docs/product/workspace/performance#data-model-compilation) to understand how the data model compilation impacts the overall performance of your deployment.

I hope this helps!

**Comment by igorlukanin** [comment]
@drewschaeffer Did the advice above help?

**Comment by batflarrow** [comment]
> Finally, I have no insight into whether you use multitenancy or not. Cube maintains a separate compiled data model for each tenant, and it would recompile it for each tenant when schema version changes.

Are you saying that when the schema version changes it will recompile all the tenants instead of the `context_to_app_id` for which the request came in?

**Comment by igorlukanin** [comment]
@batflarrow No. What I meant is that, when multitenancy is used, compiling the data model for one tenant does not remove the need to compile it for another.

**Comment by batflarrow** [comment]
> @batflarrow No. What I meant is that, when multitenancy is used, compiling the data model for one tenant does not remove the need to compile it for another.

Fair enough, so if the schema version for a single tenant changes, cube only re-compiles the schema for that account?

Also sorry for the digression, but what happens for the static list of tenants returned from scheduledRefreshContexts. Are only pre-agregations refreshed or it also refreshes the schema files and other cached content.
We were using multi tenancy in our implementation of cube so just wanted to be sure of these behaviours.
